libmamba/include/mamba/util/graph.hpp

```diff
@@ -443,127 +456,250 @@ namespace mamba::util
+        // When we explore a node with no outgoing edge, we know it is a leaf that is also a
+        // descendent of source.
+        // The pre or post order used in the node has no importance.
+        dfs_preorder_nodes_for_each_id(
+            derived_cast(),
+            [&](node_id n)
             {
-                if (g.out_degree(n) == 0)
+                if (out_degree(n) == 0)
                 {
-                    m_func(n);
+                    func(n);
                 }
-            }
-        };
-        auto visitor = LeafVisitor(func);
-        depth_first_search(visitor, source);
+            },
+            source
+        );
         return func;
     }
 
     template <typename N, typename G>
     template <typename UnaryFunc>
     UnaryFunc DiGraphBase<N, G>::for_each_root_id_from(node_id source, UnaryFunc func) const
     {
-        struct RootVisitor : default_visitor<derived_t>
-        {
-            UnaryFunc& m_func;
-
-            RootVisitor(UnaryFunc& func)
-                : m_func{ func }
+        // Explore in reverse (going in the opposite direction of the edges the directed graph
+        // starting with the given source node.
+        // When we explore a node with no incoming edge, we know it is a root that is also an
+        // ascendent of source.
+        // The pre or post order used in the node has no importance.
+        dfs_preorder_nodes_for_each_id(
+            derived_cast(),
+            [&](node_id n)
             {
-            }
+                if (in_degree(n) == 0)
+                {
+                    func(n);
+                }
+            },
+            source,
+            /* reverse= */ true
+        );
+        return func;
+    }
+
+    template <typename N, typename G>
+    auto DiGraphBase<N, G>::number_of_node_id() const noexcept -> node_id
+    {
+        // Not number_of_nodes because due to remove nodes it may be larger
+        return m_successors.size();
+    }
+
+    template <typename N, typename G>
+    auto DiGraphBase<N, G>::derived_cast() -> derived_t&
+    {
+        return static_cast<derived_t&>(*this);
+    }
+
+    template <typename N, typename G>
+    auto DiGraphBase<N, G>::derived_cast() const -> const derived_t&
+    {
+        return static_cast<const derived_t&>(*this);
+    }
+
+    /*******************************
+     *  Algorithms implementation  *
+     *******************************/
+
+    namespace detail
+    {
+        enum struct Visited
+        {
+            yes,
+            ongoing,
+            no
+        };
 
-            void start_node(node_id n, const derived_t& g)
+        template <typename Graph, typename Visitor>
+        void dfs_raw_impl(
+            const Graph& graph,
+            Visitor&& visitor,
+            typename Graph::node_id start,
+            std::vector<Visited>& status,
+            const typename Graph::adjacency_list& adjacency
+        )
+        {
+            status[start] = Visited::ongoing;
```

Comment:
Is it safe to do this instead of using `.at` here?

Reply:
If it's required that `start` is in boundaries of `status` size, maybe just add an assertion?

Reply:
I think that's a contract of the functions and the asserts are enough checks yeah.